### DISCUSSION TITLE
Units in SOFA and vibration problems

### DISCUSSION THREAD

**Discussion by Abdelrahman-Alkhodary** [SofaDefrost/SoftRobots#135] [Help - Programming with SoftRobots] [answered]
Hello everyone

I have been trying to create an arm inside SOFA that's controlled by four cables. I posted an issue before of the arm disappearing after starting the animation and the problem is solved, at that time I started by creating the arm with one cable then I added three more cables. My issue now is when I control the arm, I found that the arm breaks and vibrates with no error from SOFA. Here's a video that shows the error
https://youtu.be/0iPeY_rgTeU

and my code is here
https://github.com/Abdelrahman-Alkhodary/SoftArm-SOFA

 I have two more questions, first what are the units that SOFA use for example in cable controller in the following examble
https://github.com/SofaDefrost/SoftRobots/blob/master/docs/sofapython3/thematicalDocs/T4-DirectActuation/Actuators/Cable/CableForce.py

the value of the force is increased by 1000. The arm I'm trying to simulate should be actuated by cables with maxForce = 15 Newtons. My other question is in the example when I press the plus key, the finger moves step by step as much as I press the key but when I press the minus key the finger reset to its rest position suddenly not step by step as what happens when you press the plus key. Do you know what the solution could be for this as it also happened with the arm I'm trying to simulate?

**Comment by EulalieCoevoet** [comment, marked as answer]
Hi @Abdelrahman-Alkhodary, 

For the vibration problem, by watching the video, I can see that the deformation at the extremity of the arm is too much at some point, making some tetrahedrons "turn over". I'm familiar with the phenomenon, but can't find any documentation about that.  

For the units, **you are the one who decides what 1 means**. It's not a setting in SOFA per se, but a choice that you make when entering the parameters such as the mass, the gravity, and Young's modulus. For instance, if your mesh is in mm, then your gravity should be `9810`, because `9.81m.s-2` is `9810mm.s-2`. Now if you enter the mass in kg, and you want the Young modulus to be `5e-3GPa ->  5e6Pa -> 5e6kg.m-1.s-2 -> 5e3kg.mm-1.s-2`, you should enter `5e3` then. In the same "context", `1000` for a force means `1000kg.mm.s-2` so `1N`. 

For your last problem, the keyboard control works fine for me. Is it possible that you've changed the files? I should say that I'm on the master branch of SoftRobots.

Hope this helps.

**Reply by Abdelrahman-Alkhodary** [reply]
So, for the vibration problem do you think the problem could be in the mesh file? or how SOFA handles the mesh file?

And thanks for explaining the units I really appreciate it.

For the keys problem, I found it's from the keyboard I'm running SOFA on a laptop with no keypad so I'm accessing the + and - keys with the shift key when I changed to other keyboard keys it's working fine.

**Reply by EulalieCoevoet** [reply]
I've looked at the video again and it seems to me that you are applying to much deformation in a region where you don't have enough points/elements. You can try to change your mesh by adding more points at the extremity. 
Also we can see the cable going out of the structure, so I would add some _passing points_ too. 
To be clear, it's not that there is a problem with your mesh, but there is a problem with what you are trying to do with the mesh you have.

**Reply by EulalieCoevoet** [reply]
@Abdelrahman-Alkhodary any update? should we mark this discussion as answered?

**Reply by Abdelrahman-Alkhodary** [reply]
Hello @EulalieCoevoet actually I did not have time to try it yet, but we can mark it as answered, and if there's any new problem, I won't hesitate to post them :)
